packages/wallet-sdk/src/core/communicator/Communicator.test.ts

```diff
@@ -97,4 +98,122 @@ describe('Communicator', () => {
   });
+
+  describe('postRPCRequest', () => {
+    const request1 = { id: 'mock-request-id-1-1', data: {} } as unknown as RPCRequestMessage;
+    const request2 = { id: 'mock-request-id-2-2', data: {} } as unknown as RPCRequestMessage;
+    const simulateAsyncCall = () => new Promise((resolve) => setTimeout(resolve, 10));
+
+    let postMessageSpy: jest.SpyInstance;
+    let disconnectSpy: jest.SpyInstance;
+
+    beforeEach(() => {
+      queueMessageEvent(popupLoadedMessage);
+
+      disconnectSpy = jest.spyOn(communicator, 'disconnect');
+      postMessageSpy = jest
+        .spyOn(communicator, 'postMessage')
+        .mockImplementation(async (message: Message) => {
+          await simulateAsyncCall();
+          queueMessageEvent({
+            data: {
+              requestId: message.id,
+            },
+          });
+        });
+    });
+
+    describe('single request', () => {
+      it('should call disconnect immediately when single request resolves', async () => {
+        const response = await communicator.postRPCRequest(request1);
+        expect(response.requestId).toBe(request1.id);
+        expect(postMessageSpy).toHaveBeenCalledWith(request1);
+        expect(disconnectSpy).not.toHaveBeenCalled();
+
+        return simulateAsyncCall().then(() => {
+          expect(disconnectSpy).toHaveBeenCalledTimes(1);
+        });
+      });
+
+      it('should call disconnect immediately when single request rejects', async () => {
+        postMessageSpy = jest
+          .spyOn(communicator, 'postMessage')
+          .mockRejectedValue(new Error('mock error'));
+
+        await expect(communicator.postRPCRequest(request1)).rejects.toThrow('mock error');
+        expect(postMessageSpy).toHaveBeenCalledWith(request1);
+        expect(communicator.disconnect).toHaveBeenCalled();
+      });
+
+      it('should reject if popup disconnects before posting request', async () => {
+        const promise = communicator.postRPCRequest(request1);
+        expect(postMessageSpy).not.toHaveBeenCalledWith(request1);
+        expect(communicator.disconnect).not.toHaveBeenCalled();
+
+        // close the popup
+        communicator.disconnect();
+
+        await expect(promise).rejects.toThrow('Request cancelled before sending');
+      });
+
+      it('should reject if popup disconnects while waiting for response', async () => {
+        // no dispatchMessageEvent simulation so that the request is never resolved
+        postMessageSpy = jest.spyOn(communicator, 'postMessage').mockResolvedValue();
+
+        const promise = communicator.postRPCRequest(request1);
+        expect(postMessageSpy).not.toHaveBeenCalledWith(request1);
+        expect(communicator.disconnect).not.toHaveBeenCalled();
+
+        // close the popup after a delay
+        setTimeout(() => communicator.disconnect(), 100);
+
+        await expect(promise).rejects.toThrow('Request cancelled before response');
```

Comment:
this happens if popup closes after request promise starts execution, but before receiving response.
e.g. user manually closes the popup